plutus-ledger-api/src/PlutusLedgerApi/V3/Contexts.hs

```diff
@@ -0,0 +1,577 @@
+  | -- | Register and delegate staking credential to a Delegatee in one certificate. Noter that
+    -- deposit is mandatory.
+    TxCertRegDeleg V2.StakingCredential Delegatee V2.Value
+  | -- | Register a DRep with mandatory deposit value and an optional Anchor
+    TxCertRegDRep DRepHash V2.Value (Haskell.Maybe Anchor)
+  | -- | Update DRep's optional Anchor
+    TxCertUpdateDRep DRepHash (Haskell.Maybe Anchor)
+  | -- | UnRegister a DRep with mandatory refund value
+    TxCertUnRegDRep DRepHash V2.Value
+  | -- | Authorize a Hot credential for a specific Committee member's cold credential
+    TxCertAuthHotCommittee ColdCommitteeHash HotCommitteeHash
+  | TxCertResignColdCommittee ColdCommitteeHash
+  deriving stock (Generic, Haskell.Show, Haskell.Eq)
+  deriving (Pretty) via (PrettyShow TxCert)
+
+instance PlutusTx.Eq TxCert where
+  {-# INLINEABLE (==) #-}
+  TxCertRegStaking a b == TxCertRegStaking a' b' =
+    a PlutusTx.== a' PlutusTx.&& b PlutusTx.== b'
+  TxCertUnRegStaking a b == TxCertUnRegStaking a' b' =
+    a PlutusTx.== a' PlutusTx.&& b PlutusTx.== b'
+  TxCertDelegStaking a b == TxCertDelegStaking a' b' =
+    a PlutusTx.== a' PlutusTx.&& b PlutusTx.== b'
+  TxCertRegDeleg a b c == TxCertRegDeleg a' b' c' =
+    a PlutusTx.== a' PlutusTx.&& b PlutusTx.== b' PlutusTx.&& c PlutusTx.== c'
+  TxCertRegDRep a b c == TxCertRegDRep a' b' c' =
+    a PlutusTx.== a' PlutusTx.&& b PlutusTx.== b' PlutusTx.&& c PlutusTx.== c'
+  TxCertUpdateDRep a b == TxCertUpdateDRep a' b' =
+    a PlutusTx.== a' PlutusTx.&& b PlutusTx.== b'
+  TxCertUnRegDRep a b == TxCertUnRegDRep a' b' =
+    a PlutusTx.== a' PlutusTx.&& b PlutusTx.== b'
+  TxCertAuthHotCommittee a b == TxCertAuthHotCommittee a' b' =
+    a PlutusTx.== a' PlutusTx.&& b PlutusTx.== b'
+  TxCertResignColdCommittee a == TxCertResignColdCommittee a' =
+    a PlutusTx.== a'
+  _ == _ = Haskell.False
+
+data Voter
+  = CommitteeVoter !HotCommitteeHash
+  | DRepVoter DRepHash
+  | StakePoolVoter V2.PubKeyHash
+  deriving stock (Generic, Haskell.Show, Haskell.Eq)
+  deriving (Pretty) via (PrettyShow Voter)
+
+instance PlutusTx.Eq Voter where
+  {-# INLINEABLE (==) #-}
+  CommitteeVoter a == CommitteeVoter a' =
+    a PlutusTx.== a'
+  DRepVoter a == DRepVoter a' =
+    a PlutusTx.== a'
+  StakePoolVoter a == StakePoolVoter a' =
+    a PlutusTx.== a'
+  _ == _ = Haskell.False
+
+data Vote
+  = VoteNo
+  | VoteYes
+  | Abstain
+  deriving stock (Generic, Haskell.Show, Haskell.Eq)
+  deriving (Pretty) via (PrettyShow Vote)
+
+instance PlutusTx.Eq Vote where
+  {-# INLINEABLE (==) #-}
+  VoteNo == VoteNo   = Haskell.True
+  VoteYes == VoteYes = Haskell.True
+  Abstain == Abstain = Haskell.True
+  _ == _             = Haskell.False
+
+-- | Similar to TxOutRef, but for GovActions
+data GovernanceActionId = GovernanceActionId
+  { gaidTxId        :: V2.TxId
+  , gaidGovActionIx :: Haskell.Integer
+  }
+  deriving stock (Generic, Haskell.Show, Haskell.Eq)
+
+instance Pretty GovernanceActionId where
```

Comment:
ditto. generally unclear when you're picking one or the other. applies later too